core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -155,19 +193,34 @@
                 }
                 curr = curr.next;
             }
-            return false;
         }
+        return ownAttr ? false : hasAttrInRoot(key);
     }
 
-    /**
-     * Returns all {@link Attribute}s this map contains.
-     */
+    private <T> boolean hasAttrInRoot(AttributeKey<T> key) {
+        if (root != null) {
+            return root.hasAttr(key);
+        }
+        return false;
+    }
+
+    @Override
     public Iterator<Attribute<?>> attrs() {
         final AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
         if (attributes == null) {
-            return Collections.emptyIterator();
+            return root != null ? root.attrs() : Collections.emptyIterator();
         }
 
+        final Iterator<Attribute<?>> ownAttrsIt = new IteratorImpl(attributes);
+        return root != null ? new ConcatenatedIteratorImpl(ownAttrsIt, root.attrs())
+                            : ownAttrsIt;
```

Comment:
I guess this is not fixed yet, right? Please add a test case for this as well.

```java
AttributeMap root = ...;
root.attr(A).set(1);
root.attr(B).set(2);
AttributeMap attrs = ...;
assert attrs.isEmpty(); // attrs.iterator() will return `root.attrs()` because attrs has no attributes yet.
for (Attribute a : attrs) {
    a.set(a.get() * 10);
}
assert root.attr(A).get() == 1;
assert root.attr(B).get() == 2;
assert attrs.attr(A).get() == 10;
assert attrs.attr(B).get() == 20;
```

Reply:
Do you think https://github.com/line/armeria/pull/2322/files#diff-2bd4763eb18d444219c2be4de4da9911R292 is not sufficient? I will update it then.

Reply:
It's not enough. You return `root.attrs()` above.

Reply:
Oops Completely missed that. 😅